src/cpp/_pythia8.cpp

```diff
@@ -103,6 +104,28 @@ py::array_t<int> event_array_children(Event &event)
     return py::array_t<int>(shape, strides, ptr1);
 }
 
+// refills "event" stack with particles 
+// having pid and energy
+void refill_stack(Event &event,
+                    const ParticleData &pd,
+                    py::array_t<int> &pid,
+                    py::array_t<double> &energy) 
+{
+    int id = 0;
+    double en = 0.;
+    double mass = 0.;
+    double pz = 0.;
+    
+    event.reset();
+    for (int i = 0; i != pid.size(); ++i) {
+        id = pid.at(i);
+        en = energy.at(i);
+        mass = pd.findParticle(id)->m0();
+        pz = std::sqrt((en - mass)*(en + mass));
+        event.append(id, 91, 0, 0, 0., 0., pz, en, mass);
+    }    
```

Comment:
This is not an efficient way to iterate over a numpy array. You should request the raw memory buffer from the array and then iterate over that. It should be even faster.

Reply:
The simplest way is to call `int* pid_raw = pid.data();` to get the raw buffer.

Reply:
Actually, no, don't use this, because this only works if `py::array_t<int>` is a contiguous c-style array. This is not guaranteed in Chromo. Please read https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html#arrays for more information. You should use this code instead: https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html#direct-access

Reply:
The user should also have the option to set another status code than 91.

Reply:
And mass should also come from the user, not from ParticleData.

Reply:
* To iterate in an efficient way I used `py::array_t.unchecked<1>()`
* To append efficiently the underlying `Event.enty` vector is accessed for size reservation. 
* 2 version of `fill_event` are added. One to append pid, status, px, py, pz, energy, mass and another is the initial version to append pid and energy (it is the variant that I use, as I don't need momenta).